flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java

```diff
@@ -109,23 +125,55 @@ public void init() throws Exception {
                         this::getAsyncCheckpointStartDelayNanos);
     }
 
+    @Override
+    protected void processInput(Controller controller) throws Exception {
+        if (!isExternallyInducedSource || triggeredCheckpoints.isEmpty()) {
```

Comment:
Doesn't it cause hot looping? It also adds an additional condition on the hot path, doesn't it?

Would it be possible to `suspendDefaultAction` when we add something to the `triggeredCheckpoints` and unfreeze it once the `triggeredCheckpoints` are empty?

Reply:
Yes, I wanted to avoid it for complexity but I guess, we need to do it. That probably means tinkering with the avail future.

Reply:
I beefed `StreamTaskExternallyInducedSourceInput` a bit to not impact the hot-path of normal sources. PTAL.